clippy_lints/src/direct_recursion.rs

```diff
@@ -0,0 +1,44 @@
+use clippy_utils::diagnostics::span_lint;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// ### Why restrict this?
+    ///
```

Comment:
For the final draft of the documentation, you could mention that this lint could/should be coupled with [`#![deny(allow_attributes_without_reason)]`](https://rust-lang.github.io/rust-clippy/master/index.html#allow_attributes_without_reason)